### PR TITLE
Suggest calling the instance method of the same name when method not found

### LINKED ISSUE #103474: When a function name within a method is not found, suggest calling the method of the same name
Given the following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a8f513c3ae63f4416c7fcc1e3686a17c)):

```rust
struct S { }
impl S {
    fn first(&self) { }
    fn second(&self) { first() }
}
```

The current output is:

```
error[[E0425]](https://doc.rust-lang.org/nightly/error-index.html#E0425): cannot find function `first` in this scope
 --> src/lib.rs:4:24
  |
4 |     fn second(&self) { first() }
  |                        ^^^^^ not found in this scope
```

Ideally the output should mention the fact that a method with this name was found on the same struct. And perhaps suggest calling the method on `self`, especially if the self-param types are compatible (e.g. both take an immutable `&self` reference).

```
note: A method named `first` was found on `S`.
help:
  |
4 |     fn second(&self) { self.first() }
  |                        +++++   specify which instance of `S` to call `first` on
```

### Motivation

This may be a common error for programmers coming from C++, because C++ does not require specifying the instance for method-to-method calls.

<details><summary>C++ example</summary>
<p>

```C++
#include <iostream>
using namespace std;

class C {
    public:
        int jj() { return 1; }
        int kk() { return 5 + jj(); }
};


int main() {
    C *c = new C();
    printf("%d", c->kk());
}
```

</p>
</details>

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #103474

### CODE REVIEW

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 1]
```diff
@@ -381,11 +381,13 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
     }
 
     fn suggest_self_or_self_ref(&mut self, err: &mut Diagnostic, path: &[Segment], span: Span) {
-        let is_assoc_fn = self.self_type_is_available();
+        if !self.self_type_is_available() {
```
Comment (compiler-errors) [inline compiler/rustc_resolve/src/late/diagnostics.rs:384]: This probably causes a regression in the suggestions for the code:
```rust
struct Foo {
    i: i32,
}

impl Foo {
    fn needs_self() {
        this.i
    }
}
```

Which currently looks like:
```
error[[E0425]](https://doc.rust-lang.org/stable/error-index.html#E0425): cannot find value `this` in this scope
 --> src/lib.rs:7:9
  |
7 |         this.i
  |         ^^^^ not found in this scope
  |
help: you might have meant to use `self` here instead
  |
7 |         self.i
  |         ~~~~
help: if you meant to use `self`, you are also missing a `self` receiver argument
  |
6 |     fn needs_self(&self) {
  |                   +++++
```

Can you check that it still works? Also useful to commit a test for it.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 2]
```diff
@@ -213,26 +213,26 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
             let (mod_prefix, mod_str, suggestion) = if path.len() == 1 {
                 debug!(?self.diagnostic_metadata.current_impl_items);
                 debug!(?self.diagnostic_metadata.current_function);
-                let suggestion = if let Some(items) = self.diagnostic_metadata.current_impl_items
+                let suggestion = if self.current_trait_ref.is_none()
```
Comment (compiler-errors) [inline compiler/rustc_resolve/src/late/diagnostics.rs:216]: Why is this extra check needed now?

Comment (chenyukang) [reply compiler/rustc_resolve/src/late/diagnostics.rs:216]: We had this check before this PR, I changed the order of these "&& let" for make it more readable.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 1 above)
Comment (chenyukang) [reply compiler/rustc_resolve/src/late/diagnostics.rs:384]: This testcase still works, I didn't change the meaning of `suggest_self_or_self_ref`, I moved the check from L388 to beginning of this function, removed a variable `is_assoc_fn`.
 
I will add this code into tests.

**Pull request merged**